SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestCrossServerRPC/SpatialTestCrossServerRPC.cpp

```diff
@@ -0,0 +1,114 @@
+ *  - Each server sends RPCs to all the cubes that are not under his authority.
+ * - Test
+ *  - Server 1 checks that all the expected RPCs were received by all cubes.
+ * - Clean-up
+ *  - The previously spawned cubes are destroyed.
+ */
+
+ASpatialTestCrossServerRPC::ASpatialTestCrossServerRPC()
+	: Super()
+{
+	Author = "Andrei";
+	Description = TEXT("Test CrossServer RPCs");
+}
+
+void ASpatialTestCrossServerRPC::BeginPlay()
+{
+	Super::BeginPlay();
+
+	TArray<FVector> CubesLocations;
+	CubesLocations.Add(FVector(250.0f, 250.0f, 75.0f));
+	CubesLocations.Add(FVector(250.0f, -250.0f, 75.0f));
+	CubesLocations.Add(FVector(-250.0f, -250.0f, 75.0f));
+	CubesLocations.Add(FVector(-250.0f, 250.0f, 75.0f));
+
+	for (int i = 1; i <= 4; ++i)
+	{
+		// Each server spawns a cube
+		AddStep(TEXT("ServerSetupStep"), FWorkerDefinition::Server(i), nullptr, [this, i, CubesLocations](ASpatialFunctionalTest* NetTest)
+			{
+				ACrossServerRPCCube* TestCube = GetWorld()->SpawnActor<ACrossServerRPCCube>(CubesLocations[i-1], FRotator::ZeroRotator, FActorSpawnParameters());
+				RegisterAutoDestroyActor(TestCube);
+
+				FinishStep();
+			});
+	}
+
+	// Each server sends an RPC to all cubes that it is NOT authoritive over.
+	AddStep(TEXT("ServerSendRPCs"), FWorkerDefinition::AllServers, [this, CubesLocations](ASpatialFunctionalTest* NetTest)
+		{
+			// Make sure that all cubes were spawned and are visible to all servers before trying to send the RPCs.
+			TArray<AActor*> TestCubes;
+			UGameplayStatics::GetAllActorsOfClass(GetWorld(), ACrossServerRPCCube::StaticClass(), TestCubes);
+
+			return TestCubes.Num() == CubesLocations.Num();
+		}, [this](ASpatialFunctionalTest* NetTest)
+		{
+			TArray<AActor*> TestCubes;
+			UGameplayStatics::GetAllActorsOfClass(GetWorld(), ACrossServerRPCCube::StaticClass(), TestCubes);
+
+			for (AActor* CubeWithAuthority : TestCubes)
+			{
+				if (CubeWithAuthority->HasAuthority())
+				{
+					for (AActor* OtherActor : TestCubes)
+					{
+						if (CubeWithAuthority != OtherActor)
+						{
+							ACrossServerRPCCube* OtherCube = Cast<ACrossServerRPCCube>(OtherActor);
+							OtherCube->CrossServerTestRPC(GetLocalFlowController()->GetWorkerDefinition().Id);
+						}
+					}
+
+					break;
+				}
+			}
+
+			FinishStep();
+		});
+
+	// Server 1 checks if all cubes received the expected number of RPCs.
+	AddStep(TEXT("Server1CheckRPCs"), FWorkerDefinition::Server(1), nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime)
+		{
+			TArray<AActor*> TestCubes;
+			UGameplayStatics::GetAllActorsOfClass(GetWorld(), ACrossServerRPCCube::StaticClass(), TestCubes);
+
+			int TotalNumberOfCubes = TestCubes.Num();
+			int CorrectCubes = 0;
+
+			for (int i = 0; i < TotalNumberOfCubes; ++i)
+			{
+				ACrossServerRPCCube* CurrentCube = Cast<ACrossServerRPCCube>(TestCubes[i]);
+
+				int ReceivedRPCS = CurrentCube->ReceivedCrossServerRPCS.Num();
+
+				if (ReceivedRPCS == TotalNumberOfCubes - 1)
```

Comment:
The assert looks useless, since you only test it in a similar conditional.

Reply:
(otherwise, I assume it times out after 5 sec, which could be all it needs)

Reply:
Yes that is correct, just thought to keep it around for some debugging information, since the test might fail but you wouldn't know for sure for which cube it failed, it doesn't bring much information so could be deleted if you think it's not worth having. 

Regarding breaking the test, I've managed to break it yes, but it a kind of 'artificial' way, by not processing RPCs sent by one of the cubes.